### PR TITLE
Antithetic estimator.

### PR DESCRIPTION
**Pull request by jamie-mcg**

### CODE REVIEW

File: src/optim/FishLeg/fishleg.py [hunk 1]
```diff
@@ -97,6 +97,8 @@ def __init__(
         damping: float = 5e-1,
         update_aux_every: int = 10,
         warmup_steps: int = 0,
+        method: str = "rank-1",
+        eps: float = 1e-4,
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:101]: not a big fan of this... maybe we could have a dictionary for method_parameters and inside the method we check if it contains the necessary parameters? ... not sure.. I dont like that this eps is only used for the antithetic

File: src/optim/FishLeg/fishleg.py [hunk 2]
```diff
@@ -278,34 +283,99 @@ def update_aux(self) -> None:
...
-                nat_grads = module.Qv(layer_grads)
+            aux_loss /= 1.0 + torch.norm(plus_loss.detach() + minus_loss.detach())
 
-                for n, (name, param) in enumerate(module.named_parameters()):
-                    if "fishleg_aux" not in name:
-                        nat_grad = nat_grads[n]
-                        sample_grad = grads[p_idx]
-                        true_grad = layer_grads[n]
-                        p_idx += 1
+        else:
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:350]: should be elif, with rank-1. and the else at the end should raise a not-implemented error

File: src/optim/FishLeg/fishleg.py [hunk 3]
```diff
@@ -278,34 +283,99 @@ def update_aux(self) -> None:
...
-            if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
-                layer_grads = []
-                for name, param in module.named_parameters():
-                    if "fishleg_aux" not in name:
-                        layer_grads.append(param.grad.data / u_norm)
+        if self.method == "antithetic":
+            nat_grads = []
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:292]: would be nicer here (and in the following occurrences) if we maybe just check that it is type FishModule? what do you think?

File: src/optim/FishLeg/fishleg.py [hunk 4]
```diff
@@ -97,6 +97,8 @@ def __init__(
         damping: float = 5e-1,
         update_aux_every: int = 10,
         warmup_steps: int = 0,
+        method: str = "rank-1",
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:100]: not related to this line, but two lines down for the device variable (not sure why but I could not write this comment there). anyway: 
if we introduce the stack overflow suggestion that I have sent you @jamie, maybe we might be able to avoid passing the device to the optimizer? let's have a chat about this

File: src/optim/FishLeg/fishleg.py [hunk 5]
```diff
@@ -278,34 +283,99 @@ def update_aux(self) -> None:
...
+            minus_loss = self.likelihood.nll(pred_y, samples_y)
+
+            # Restore parameters
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
+                            param.data += self.eps * nat_grads[p_idx]
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:343]: We should also have a chat about this. We need to choose between memory and computational speed. I feel a bit torn.

File: src/optim/FishLeg/fishleg.py (hunk 4 above)
Comment (jamie-mcg) [reply src/optim/FishLeg/fishleg.py:100]: Will take a look after lunch and see about this, I think this sounds good though.

File: src/optim/FishLeg/fishleg.py (hunk 5 above)
Comment (jamie-mcg) [reply src/optim/FishLeg/fishleg.py:343]: Sure, yes I was also thinking this. I went this way first since it is simpler and also the memory capacity is a hard limit, whereas time is less restrictive. I think having an option for parallelising this could be good though. There may also be other ways to structure the computation which may help.

File: src/optim/FishLeg/fishleg.py [hunk 6]
```diff
@@ -278,34 +283,99 @@ def update_aux(self) -> None:
...
-                    if "fishleg_aux" not in name:
-                        layer_grads.append(param.grad.data / u_norm)
+        if self.method == "antithetic":
+            nat_grads = []
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:295]: can we have flags for the parameters as an attribute? or do we need to have a new class definition? I am not a very big fan of replying on the name of the parameter to check if it is a auxiliary parameter or not. I upderstand it is not high priority at all to improve this, and we can just go on as it is. We use this condition so often that it might be even worth having a new function that filters this out in advance for polishness purposes. But again. Not priority. We can have a chat about this anyway and decide together what we think might be worth touching or not

File: src/optim/FishLeg/fishleg.py [hunk 7]
```diff
@@ -278,34 +283,99 @@ def update_aux(self) -> None:
...
+            pred_y = self.model(data_x)
+
+            minus_loss = self.likelihood.nll(pred_y, samples_y)
+
+            # Restore parameters
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:341]: we could also have just a function that does the same for each of these, but we only give a different scalar pre-multiplier for each. I am not sure if it would improve or reduce readibility though, we can have a chat about it if you want

File: src/optim/FishLeg/fishleg.py [hunk 8]
```diff
@@ -278,34 +283,99 @@ def update_aux(self) -> None:
...
+                    qv, qv_b = module.Qv(layer_grads)
+                    aux_loss -= torch.sum(grads[p_idx] * qv)
+                    aux_loss -= torch.sum(grads[p_idx + 1] * qv_b)
+
+                    aux_loss += damping / 2 * torch.sum(qv**2)
+                    aux_loss += damping / 2 * torch.sum(qv_b**2)
+
+                    nat_grads.append(qv)
+                    nat_grads.append(qv_b)
+                    p_idx += 2
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:307]: solution compatible with other layer types:

layer_nat_grads = module.Qv(layer_grads)
for i, gr in enumerate(layer_nat_grads):
    aux_loss -= torch.sum(grads[p_idx + i] * gr)
    aux_loss += damping / 2 * torch.sum(gr**2)
    nat_grads.append(gr)
p_idx += i

File: src/optim/FishLeg/fishleg.py [hunk 9]
```diff
@@ -285,11 +261,16 @@ def update_aux(self) -> None:
 
         aux_loss = 0
 
-        if self.method == "antithetic":
+        if method == "antithetic":
+            try:
+                eps = method_kwargs["eps"]
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:266]: eps = method_kwargs["eps"] if "eps" in method_kwargs else 1e-4

Comment (federicamtk) [reply src/optim/FishLeg/fishleg.py:266]: this should work better. the try, except seems unnecessary to me here. Should be just an if condition

File: src/optim/FishLeg/fishleg.py [hunk 10]
```diff
@@ -258,11 +233,12 @@ def update_aux(self) -> None:
         """
         self.aux_opt.zero_grad()
 
-        data_x, data_y = next(iter(self.aux_dataloader))
-        data_x, data_y = data_x.to(self.device), data_y.to(self.device)
+        data_x, _ = next(iter(self.aux_dataloader))
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:236]: excellent, just  remember to change the dataloader call in the other scripts to include the collate_fn() that we talked about

File: src/optim/FishLeg/fishleg.py [hunk 11]
```diff
@@ -395,15 +362,21 @@ def update_aux(self) -> None:
...
         return aux_loss.item()
 
-    # TODO: What do we need here? - This method is taken from Adam, to discuss!
+    def _augment_params_by(self, eps: float, nat_grads: List):
+        p_idx = 0
+        for module in self.model.modules():
+            if isinstance(module, FishModule):
+                for name, param in module.named_parameters():
+                    if "fishleg_aux" not in name:
+                        param.data -= 2 * eps * nat_grads[p_idx]
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:371]: This function looks good but I think it might be worth defining it as small function snipped inside the update_aux fnction (instead of defining it as a self function); if you did that you would not need to pass the nat_grads either as input and also we are not going to use it anywhere else anyway. Anyway it is just a small detail, here it is fine too. However, according to how you used this function above you should not be using the form 
```
param.data -= 2 * eps * nat_grads[p_idx]
```
but you should be using the form 
```
param.data +=  eps * nat_grads[p_idx]
```
otherwise I think you are introducing a bug, from what I can see

File: src/optim/FishLeg/fishleg.py [hunk 12]
```diff
@@ -240,92 +174,108 @@ def update_aux(self) -> None:
...
-        )
-
-        u2 = 0.0
-        for group in self.param_groups:
-            for params in group["params"]:
-                u2 = u2 + torch.sum(params.grad * params.grad)
-        u_norm = torch.sqrt(u2)
+        g2 = 0.0
+        for g in group["params"]:
+            g2 += torch.sum(g**2)
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:183]: isnt the name g confusing? as we are not using gradients here?

File: src/optim/FishLeg/fishleg.py [hunk 13]
```diff
@@ -240,92 +174,108 @@ def update_aux(self) -> None:
...
-                        if "fishleg_aux" not in name:
-                            layer_grads.append(param.grad.data / u_norm)
-
-                    nat_grads = module.Qv(layer_grads)
-
                     for n, (name, param) in enumerate(module.named_parameters()):
                         if "fishleg_aux" not in name:
-                            nat_grad = nat_grads[n]
-                            sample_grad = grads[p_idx]
+                            nat_grad = v_model[n]
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:277]: isnt it the other way around with the indexing?

File: src/optim/FishLeg/fishleg.py [hunk 14]
```diff
@@ -240,92 +174,108 @@ def update_aux(self) -> None:
...
-            allow_unused=True,
-        )
-
-        u2 = 0.0
-        for group in self.param_groups:
-            for params in group["params"]:
-                u2 = u2 + torch.sum(params.grad * params.grad)
-        u_norm = torch.sqrt(u2)
+        g2 = 0.0
+        for g in group["params"]:
```
Comment (federicamtk) [inline src/optim/FishLeg/fishleg.py:182]: isn't the name g confusing? they are not gradients, maybe we could keep using u?

File: src/optim/FishLeg/fishleg.py (hunk 12 above)
Comment (federicamtk) [reply src/optim/FishLeg/fishleg.py:183]: this should be params.grad, otherwise it will take the parameter value instead of the gradient

Comment (federicamtk) [reply src/optim/FishLeg/fishleg.py:183]: grad.data

Comment (federicamtk) [reply src/optim/FishLeg/fishleg.py:183]: 248 layer_grads

**Comment by federicamtk** [comment]
several bugs in fisher analysis, wrong targets and also the code breaks for any N different from 100. Issue with batch size and matrix multiplication of the input data.

**Comment by federicamtk** [comment]
further bugs for fisher anaysis, debugging in progress, to discuss.
